fileserver/fileop.go

```diff
@@ -3412,6 +3423,369 @@ func indexRawBlocks(repoID string, blockIDs []string, fileHeaders []*multipart.F
+	if err != nil {
+		return nil, &appError{err, "", http.StatusInternalServerError}
+	}
+	if status != repomgr.RepoStatusNormal && status != -1 {
+		msg := "Repo status not writable."
+		return nil, &appError{nil, msg, http.StatusBadRequest}
+	}
+
+	user, _ := repomgr.GetRepoOwner(repoID)
+
+	fsm := new(recvData)
+
+	fsm.parentDir = parentDir
+	fsm.tokenType = "upload-link"
+	fsm.repoID = repoID
+	fsm.user = user
+	fsm.rstart = -1
+	fsm.rend = -1
+	fsm.fsize = -1
+
+	ranges := r.Header.Get("Content-Range")
+	if ranges != "" {
+		parseContentRange(ranges, fsm)
+	}
+
+	return fsm, nil
+}
+
+type ShareLinkInfo struct {
+	RepoID    string `json:"repo_id"`
+	FilePath  string `json:"file_path"`
+	ParentDir string `json:"parent_dir"`
+}
+
+func queryAccessToken(token, opType string) (*ShareLinkInfo, *appError) {
+	claims := SeahubClaims{
+		time.Now().Add(time.Second * 300).Unix(),
+		true,
+		jwt.RegisteredClaims{},
+	}
+
+	jwtToken := jwt.NewWithClaims(jwt.GetSigningMethod("HS256"), &claims)
+	tokenString, err := jwtToken.SignedString([]byte(seahubPK))
+	if err != nil {
+		err := fmt.Errorf("failed to sign jwt token: %v", err)
+		return nil, &appError{err, "", http.StatusInternalServerError}
+	}
+	url := fmt.Sprintf("%s?token=%s&type=%s", seahubURL+"/share-link-info/", token, opType)
+	header := map[string][]string{
+		"Authorization": {"Token " + tokenString},
+	}
+	status, body, err := utils.HttpCommon("GET", url, header, nil)
+	if err != nil {
+		err := fmt.Errorf("failed to get share link info: %v", err)
+		return nil, &appError{err, "", http.StatusInternalServerError}
+	}
+	if status != http.StatusOK {
+		msg := "Access token not found"
+		return nil, &appError{nil, msg, http.StatusForbidden}
+	}
+
+	info := new(ShareLinkInfo)
+	err = json.Unmarshal(body, &info)
+	if err != nil {
+		err := fmt.Errorf("failed to decode share link info: %v", err)
+		return nil, &appError{err, "", http.StatusInternalServerError}
+	}
+
+	return info, nil
+}
+
+func accessLinkCB(rsp http.ResponseWriter, r *http.Request) *appError {
+	if seahubPK == "" {
+		err := fmt.Errorf("no seahub private key is configured")
+		return &appError{err, "", http.StatusNotFound}
+	}
+
+	parts := strings.Split(r.URL.Path[1:], "/")
+	if len(parts) < 2 {
+		msg := "Invalid URL"
+		return &appError{nil, msg, http.StatusBadRequest}
+	}
+	token := parts[1]
+	info, appErr := queryAccessToken(token, "file")
+	if appErr != nil {
+		return appErr
+	}
+
+	if info.FilePath == "" {
+		msg := "Invalid file_path\n"
+		return &appError{nil, msg, http.StatusBadRequest}
+	}
+
+	repoID := info.RepoID
+	filePath := normalizeUTF8Path(info.FilePath)
+	fileName := filepath.Base(filePath)
+	op := "download-link"
+
+	if _, ok := r.Header["If-Modified-Since"]; ok {
+		return &appError{nil, "", http.StatusNotModified}
+	}
+
+	now := time.Now()
+	rsp.Header().Set("Last-Modified", now.Format("Mon, 2 Jan 2006 15:04:05 GMT"))
+	rsp.Header().Set("Cache-Control", "max-age=3600")
```

Comment:
Last-Modified 应该设置为文件的最后修改时间吧。

如果请求设置了 If-Modified-Since 头部或者 If-None-Match 头部，那么需要检查文件是否有被更新。参考 https://developer.mozilla.org/en-US/docs/Web/HTTP/Caching#validation 。同时我们应该设置 Cache-Control: no-cache，让客户端先检查缓存是否已经过期。

这些检查和头部设置应该都在获取到文件元信息之后才能做的吧。C 部分也需要相应改动。